### PR TITLE
Bicep module for Function app

### PR DESCRIPTION
**Pull request by rhalloul**
## Description

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

Closes https://github.com/Azure/bicep-registry-modules/issues/259.

## Adding a new module

- [ ] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ ] I have updated the examples in README with the latest module version number.

### CODE REVIEW

File: modules/compute/function-app/main.bicep [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+  properties: {
+    perSiteScaling: perSiteScaling
+    maximumElasticWorkerCount: maximumElasticWorkerCount
+    reserved: serverOS == 'Linux'
+    targetWorkerCount: targetWorkerCount
+    targetWorkerSizeId: targetWorkerSizeId
+  }
+}
+
+@description('Defines insight for app or function app i.e sites.')
```
Comment (thomasriley) [inline modules/compute/function-app/main.bicep:187]: Small thing, but the `@description` decorator is only consumed by Bicep on parameters and outputs. On the resource blocks, it might be a little more human readable if you use a code comment instead?

**Review by thomasriley — CHANGES_REQUESTED** [review]

File: modules/compute/function-app/main.bicep [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+param branch string = ''
+
+@description('Optional. to limit to manual integration; to enable continuous integration (which configures webhooks into online repos like GitHub).')
+param isManualIntegration bool = true
+
+@description('Optional. true for a Mercurial repository; false for a Git repository.')
+param isMercurial bool = false
+
+@description('Required. Name of the storage account used by function app.')
+param storgeAccountName string
```
Comment (haack42) [inline modules/compute/function-app/main.bicep:140]: Should this not be named **storageAccount** instead of **storgeAccount**?

Comment (rhalloul) [reply modules/compute/function-app/main.bicep:140]: Thanks  Matthias

File: modules/compute/function-app/main.bicep [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  properties: {
+    subnetResourceId: subnetId
+  }
+}
+
+
+@description('The resources actual is function where code exits')
+@batchSize(1)
+resource azureFunction 'Microsoft.Web/sites/functions@2021-02-01' = [for function in functions: {
+  dependsOn: [
```
Comment (Gordonby) [inline modules/compute/function-app/main.bicep:257]: Is this dependsOn necessary?

File: modules/compute/function-app/main.bicep [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+
+@description('Defines storageAccounts for Azure Function App.')
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-06-01' existing = {
+  name: storgeAccountName
+  scope: resourceGroup(storgeAccountResourceGroup)
+}
+
+@description('Defines Application service plan.')
+resource serverfarms 'Microsoft.Web/serverfarms@2021-02-01' = {
```
Comment (Gordonby) [inline modules/compute/function-app/main.bicep:173]: Is there a reason to be using an API versions throughout?

File: modules/compute/function-app/main.bicep (hunk 2 above)
Comment (Gordonby) [inline modules/compute/function-app/main.bicep:140]: Could add maxLength decorator to prevent storage account names longer than the 24 character limit?

File: modules/compute/function-app/main.bicep [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+@description('Deploy function app from zip file.')
+resource extensions 'Microsoft.Web/sites/extensions@2022-03-01' = if (enablePackageDeploy) {
+  parent: sites
+  name: 'MSDeploy'
+  properties: {
+    packageUri: functionPackageUri
+  }
+}
+
+/*output section*/
```
Comment (Gordonby) [inline modules/compute/function-app/main.bicep:291]: Another useful output is the function apps ` defaultHostName`

**Review by Gordonby — CHANGES_REQUESTED** [review]

File: modules/compute/function-app/main.bicep (hunk 4 above)
Comment (rhalloul) [reply modules/compute/function-app/main.bicep:173]: Hosting plan is an optional requirement, but it's required if you run your function on a premium plan , it was added as dependency

File: modules/compute/function-app/README.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+- Make sure enableaInsights: true.
+- with functions of array
+
+Dependency list: 
+- Microsoft.ManagedIdentity/userAssignedIdentities
+- Microsoft.OperationalInsights/workspaces
+
+
+''')
+module test '../main.bicep' = {
```
[the remark this answers is not in the document]
Comment (rhalloul) [reply modules/compute/function-app/README.md:87]: made the changes

File: modules/compute/function-app/test/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,95 @@
+targetScope = 'subscription'
```
[the remark this answers is not in the document]
Comment (rhalloul) [reply modules/compute/function-app/test/main.test.bicep:1]: removed subscription and use resourceGroup instead

File: modules/compute/function-app/main.bicep (hunk 3 above)
Comment (rhalloul) [reply modules/compute/function-app/main.bicep:257]: Hosting plan is not required for Function app but it's required as dependency fro premium plan

File: modules/compute/function-app/main.bicep (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhalloul) [reply modules/compute/function-app/main.bicep:140]: Change the name

File: modules/compute/function-app/main.bicep (hunk 1 above)
Comment (rhalloul) [reply modules/compute/function-app/main.bicep:187]: Add description describing monitor application using log analytics if the resource is enabled.

File: modules/compute/function-app/main.bicep [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+@description('Required. Defines the name, tier, size, family and capacity of the app service plan.')
+param sku object
+
+
+@description('Optional. Kind of server OS.')
+@allowed([
+  'Windows'
+  'Linux'
+])
+
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:20]:
```suggestion
```

File: modules/compute/function-app/main.bicep [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+@description('Required. Defines the name, tier, size, family and capacity of the app service plan.')
+param sku object
+
+
+@description('Optional. Kind of server OS.')
+@allowed([
+  'Windows'
+  'Linux'
+])
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:19]:
```suggestion
@allowed([ 'Windows', 'Linux' ])
```

Comment (dciborow) [reply modules/compute/function-app/main.bicep:19]: when under 120 characters, i like to single line these arrays

File: modules/compute/function-app/main.bicep [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+
+@description('Optional. Scaling worker count.')
+param targetWorkerCount int = 0
+
+@description('Optional. The instance size of the hosting plan (small, medium, or large).')
+@allowed([
+  0
+  1
+  2
+])
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:37]:
```suggestion
@allowed([ 0, 1, 2 ])
```

File: modules/compute/function-app/main.bicep [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+param functionsWorkerRuntime string = ''
+
+@description('Optional. NodeJS version.')
+param functionsDefaultNodeversion string = '~14'
+
+@allowed([
+  'Disabled'
+  'Enabled'
+])
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:88]:
```suggestion
@allowed([ 'Disabled', 'Enabled' ])
```

File: modules/compute/function-app/main.bicep [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+@description('Optional. The network access type for accessing Application Insights ingestion. - Enabled or Disabled.')
+param publicNetworkAccessForIngestion string = 'Enabled'
+
+@allowed([
+  'Disabled'
+  'Enabled'
+])
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:95]:
```suggestion
@allowed([ 'Disabled', 'Enabled' ])
```

File: modules/compute/function-app/test/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+      family: 'Y'
+      capacity: 0
+    }
+    tags: tags
+    storageAccountName: dependencies.outputs.saAccountName
+    storgeAccountResourceGroup: resourceGroup().name
+  
+  }
+  scope: resourceGroup()
+
```
Comment (dciborow) [inline modules/compute/function-app/test/main.test.bicep:58]:
```suggestion
```

File: modules/compute/function-app/test/main.test.bicep [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  scope: resourceGroup()
+  name: 'dependencies-${uniqueString(resourceGroup().id)}'
+  params: {
+    name: name
+    location: location
+    tags: tags
+  }
+}
+
+
```
Comment (dciborow) [inline modules/compute/function-app/test/main.test.bicep:25]:
```suggestion

module test0 '../main.bicep' = {
  name: 'func-app0-${guid(name)}'
  dependsOn: [
    dependencies
  ]
  params: {
    name: 'funcapp0-${resourceGroup().name}'
    location: location
  }
  scope: resourceGroup()
}
```

Comment (dciborow) [reply modules/compute/function-app/test/main.test.bicep:25]: i like to include a test that provides the most minimum amount of input parameters.

**Review by dciborow — APPROVED** [review]

File: modules/compute/function-app/test/dependencies.bicep
Comment (dciborow) [inline modules/compute/function-app/test/dependencies.bicep:1]: nitpick: we tend to name this files like this prereq.test.bicep

**Comment by dciborow** [comment]
> Fixes: issue #259, waiting for proposal approval

Can you added "Closes #259" into the description so it automatically gets picked up?

File: modules/compute/function-app/main.bicep [hunk 15]
```diff
@@ -0,0 +1,299 @@
+/*param section*/
+@description('Required. Name for the Azure Function App.')
+param name string
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:3]:
```suggestion
@minLength(6)
@maxLength(64)
param name string = take('func${uniqueString(resourceGroup().id, subscription().id, location)}', 64)
```

Comment (dciborow) [reply modules/compute/function-app/main.bicep:3]: Tests are currently failing with `Status Message: The provided deployment name 'func-app2-267-4b15b98-20230408010517-f4e616e5fc1b4c2d813dddebe66c6d6a' has a length of '69' which exceeds the maximum length of '64'.`

This will help catch that earlier.

File: modules/compute/function-app/test/main.test.bicep [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+param name string = deployment().name
```
Comment (dciborow) [inline modules/compute/function-app/test/main.test.bicep:2]:
```suggestion
param name string = take(deployment().name, 64)
```

File: modules/compute/function-app/main.bicep [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+
+@description('Required. Location for all resources.')
+param location string
+
+@description('Optional. Tags for all resources within Azure Function App module.')
+param tags object = {}
+
+@description('Required. Defines the name, tier, size, family and capacity of the app service plan.')
+param sku object
```
Comment (dciborow) [inline modules/compute/function-app/main.bicep:12]:
```suggestion
param sku object = {
  name: 'Y1'
  tier: 'Dynamic'
  size: 'Y1'
  family: 'Y'
  capacity: 0
}
```

Comment (dciborow) [reply modules/compute/function-app/main.bicep:12]: We can include a default value here. This helps to show users how the object should be created.

**Comment by shenglol** [comment]
Sorry, I should have mentioned it in the proposal, but does it make more sense to put the module under the `web` group instead of `compute`?

**Comment by vlahane** [comment]
@dciborow @shenglol please review this one. Thanks!

File: modules/compute/function-app/main.bicep (hunk 15 above)
Comment (shenglol) [reply modules/compute/function-app/main.bicep:3]: I think adding `@maxLength(64)` makes sense. We should probably not set a default value for the parameter, though.

Comment (vlahane) [reply modules/compute/function-app/main.bicep:4]: updated @shenglol

**Comment by shenglol** [comment]
Pending on https://github.com/microsoft/mcr/pull/2520.

**Review by shenglol — APPROVED** [review]
